src/Renderer.ts

```diff
@@ -223,37 +271,40 @@ export default class Renderer {
     }
   }
 
-  setRenderTarget(renderTarget) {
+  public setRenderTarget(renderTarget: HTMLElement | string | null): void {
     if (typeof renderTarget === "string") {
-      renderTarget = document.querySelector(renderTarget);
+      renderTarget = document.querySelector(renderTarget) as HTMLElement;
     }
     this.renderTarget = renderTarget;
-    this.renderTarget.classList.add("leopard__project");
-    this.renderTarget.style.width = `${this.project.stage.width}px`;
-    this.renderTarget.style.height = `${this.project.stage.height}px`;
+    if (!renderTarget) return;
+    renderTarget.classList.add("leopard__project");
+    renderTarget.style.width = `${this.project.stage.width}px`;
+    renderTarget.style.height = `${this.project.stage.height}px`;
 
-    this.renderTarget.append(this.stage);
+    renderTarget.append(this.stage);
   }
 
   // Handles rendering of all layers (including stage, pen layer, sprites, and all clones) in proper order.
-  _renderLayers(layers, options = {}) {
-    options = Object.assign(
-      {
-        drawMode: ShaderManager.DrawModes.DEFAULT,
-        renderSpeechBubbles: true
-      },
-      options
-    );
+  private _renderLayers(
+    layers?: Set<Sprite | Stage | PenSkin>,
+    optionsIn: {
+      filter?: (layer: Sprite | Stage | PenSkin) => boolean;
+    } & Partial<RenderSpriteOptions> = {}
+  ): void {
+    const options = {
+      drawMode: ShaderManager.DrawModes.DEFAULT,
+      renderSpeechBubbles: true,
+      ...optionsIn,
+    };
```

Comment:
> We're only using the `filter` option here

I'm not sure what you mean by this. The function also accesses the `drawMode` option. I could remove the `renderSpeechBubble` de/restructuring though?

Reply:
I'm not 100% sure what I was getting at when I wrote this comment lmao, sorry.

But my (current) understanding is that `filter` isn't actually used outside of `_renderLayers` - hence its type definition there rather than on `RenderSpriteOptions`. So the `options` object we end up passing along, which is defined here... ends up being a full `RenderSpriteOptions` *with extra* options from `optionsIn` (i.e. `filter`), which inevitably goes unused.

We still need an options input to pass it to `renderSprite` and to access or provide `drawMode`. But the only reason *for* destructure `filter` from it is for legibility, when calling `_renderLayers`; and this way gives us a complicated type definition `{ filter?: ... } & ...` and we end up passing along a completely unused `filter` option in an object where it doesn't fit (`RenderSpriteOptions`).

(We don't ever destructure anything except the `filter` option, but we need to keep `renderSpeechBubbles: true` in the `options` variable because it's serving as a default, the value unless specifically overridden by `optionsIn`. Removing it would be a behavior change, defaulting to undefined instead of true.)

My rewrite would be the `_renderLayers` definition I wrote above, but with `options` renamed to `optionsIn` like you have it, and then:

```js
private _renderLayers(...): void {
  // IFF there's no cleaner way to provide defaults for drawMode
  // and renderSpeechBubbles directly in parameters
  const options = {
    drawMode: ShaderManager.DrawModes.DEFAULT,
    renderSpeechBubbles: true,
    ...optionsIn;
  };

  const shouldRestrictLayers = layers instanceof Set;
  const shouldIncludeLayer = (layer: Sprite | Stage | PenSkin): boolean =>
    !(
      // just access filter as-is instead of from options
      (shouldRestrictLayers && !layers.has(layer)) ||
      (filter && !filter(layer))
    );
  ...
}
```

Reply:
I've updated the code to pass `filter` separately.

Reply:
I think this is still a behavior change? `renderSpeechBubble`, provided to following `renderSprite` calls via `options`, is defaulting to undefined instead of true in your current code.

Reply:
I've changed `renderSprite` to only skip rendering speech bubbles if it's explicitly set to `false`.